File: android/app/src/main/java/ai/stagecraft/android/PluginConfigStore.java
package ai.stagecraft.android;

import android.content.Context;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.nio.charset.StandardCharsets;

/**
 * W6：主进程 PluginConfigStore（计划 §2.1/§5.3/§6，阶段 5；Q4 裁决）。
 *
 * 只持久化 desiredEnabled/config/quarantine/launchPlan——全部在 Core 之外，
 * Core 从未启动或已 crashed 时仍可读写（阶段 1 验收锚点）。
 * 不实现依赖拓扑/冲突规则（深度校验唯一实现在 TS plugin-bootstrap.ts，Q4 裁决），
 * 本类只做 JSON 文件原子持久化（先写临时文件再 rename）。
 *
 * 存储形状与 TS createNodeFilePluginConfigStore 对齐（formatVersion/config/enabled/
 * quarantine/launchPlan），保证桌面与 Android 同构。
 */
public final class PluginConfigStore {
    private static final int STORE_FORMAT_VERSION = 1;
    private final File file;

    public PluginConfigStore(Context context) {
        this.file = new File(context.getFilesDir(), "plugin-config-store.json");
    }

    /** 测试 seam：指定文件。 */
    public PluginConfigStore(File file) {
        this.file = file;
    }

    public File file() { return file; }

    private JSONObject load() {
        try (FileInputStream input = new FileInputStream(file)) {
            byte[] bytes = new byte[(int) file.length()];
            int read = input.read(bytes);
            String json = new String(bytes, 0, Math.max(0, read), StandardCharsets.UTF_8);
            JSONObject parsed = new JSONObject(json);
            if (parsed.optInt("formatVersion", -1) != STORE_FORMAT_VERSION) return emptyState();
            return parsed;
        } catch (Exception error) {
            // 文件不存在/损坏：空状态起步（Core 不可用时管理器必须仍可用）
            return emptyState();
        }
    }

    private static JSONObject emptyState() {
        JSONObject state = new JSONObject();
        try {
            state.put("formatVersion", STORE_FORMAT_VERSION);
            state.put("updatedAt", "1970-01-01T00:00:00.000Z");
            state.put("config", new JSONObject());
            state.put("enabled", new JSONObject());
            state.put("quarantine", new JSONArray());
            state.put("launchPlan", JSONObject.NULL);
        } catch (Exception ignored) { }
        return state;
    }

    private synchronized void persist(JSONObject state) {
        try {
            state.put("updatedAt", java.time.Instant.now().toString());
            File parent = file.getParentFile();
            if (parent != null && !parent.exists()) parent.mkdirs();
            File temporary = new File(file.getParentFile(), file.getName() + ".tmp");
            try (FileOutputStream output = new FileOutputStream(temporary)) {
                output.write(state.toString(2).getBytes(StandardCharsets.UTF_8));
            }
            if (file.exists()) file.delete();
            temporary.renameTo(file);
        } catch (Exception error) {
            GateALog.w("plugin config store persist failed: " + error);
        }
    }

    /** 读全部启用意图（id → boolean）。 */
    public synchronized JSONObject readEnabled() {
        return load().optJSONObject("enabled");
    }

    /** 写单个启用意图。 */
    public synchronized void writeEnabled(String id, boolean enabled) {
        JSONObject state = load();
        JSONObject enabledMap = state.optJSONObject("enabled");
        if (enabledMap == null) { enabledMap = new JSONObject(); try { state.put("enabled", enabledMap); } catch (Exception ignored) { } }
        try { enabledMap.put(id, enabled); } catch (Exception ignored) { }
        persist(state);
    }

    /** 读全部配置（id → config）。 */
    public synchronized JSONObject readConfig() {
        return load().optJSONObject("config");
    }

    /** 写单个插件配置。 */
    public synchronized void writeConfig(String id, JSONObject config) {
        JSONObject state = load();
        JSONObject configMap = state.optJSONObject("config");
        if (configMap == null) { configMap = new JSONObject(); try { state.put("config", configMap); } catch (Exception ignored) { } }
        try {
            if (config == null) configMap.remove(id);
            else configMap.put(id, config);
        } catch (Exception ignored) { }
        persist(state);
    }

    /** 读隔离记录（Core 上报，主进程只读展示）。 */
    public synchronized JSONArray readQuarantine() {
        return load().optJSONArray("quarantine");
    }

    /** 写隔离记录（Core bootstrap 结果经 Binder/数据面回报）。 */
    public synchronized void writeQuarantine(JSONArray records) {
        JSONObject state = load();
        try { state.put("quarantine", records == null ? new JSONArray() : records); } catch (Exception ignored) { }
        persist(state);
    }

    /** 读最近一次 launch plan（Core 重启后复用同一 plan 身份）。 */
    public synchronized JSONObject readLaunchPlan() {
        Object plan = load().opt("launchPlan");
        return plan instanceof JSONObject ? (JSONObject) plan : null;
    }

    /** 写 launch plan。 */
    public synchronized void writeLaunchPlan(JSONObject plan) {
        JSONObject state = load();
        try { state.put("launchPlan", plan == null ? JSONObject.NULL : plan); } catch (Exception ignored) { }
        persist(state);
    }
}

File: src/plugin-config-store.ts
/**
 * PluginConfigStore 实现（评审修订）：主进程配置持久化不依赖 Core——Core 从未启动或已
 * crashed 时仍可读写（阶段 1 验收锚点）。
 *
 * - InMemoryPluginConfigStore：测试与 harness。
 * - createNodeFilePluginConfigStore：桌面/Node 侧参考实现（单 JSON 文件，原子替换写）。
 * - Android 主进程的 Java 实现（PluginConfigStore.java，只持久化
 *   desiredEnabled/config 并展示隔离记录，不实现依赖拓扑规则）。
 *
 * 本模块不得 import Core runtime；规则实现只在 src/plugin-bootstrap.ts。
 */

import { mkdirSync, readFileSync, renameSync, writeFileSync } from 'node:fs'
import { dirname } from 'node:path'
import type { PluginConfigStore, PluginLaunchPlan, QuarantineRecord } from './plugin-contract.ts'

const STORE_FORMAT_VERSION = 1

interface PersistedState {
  formatVersion: number
  updatedAt: string
  config: Record<string, unknown>
  enabled: Record<string, boolean>
  quarantine: QuarantineRecord[]
  launchPlan: PluginLaunchPlan | null
}

function emptyState(): PersistedState {
  return { formatVersion: STORE_FORMAT_VERSION, updatedAt: new Date(0).toISOString(), config: {}, enabled: {}, quarantine: [], launchPlan: null }
}

export function createInMemoryPluginConfigStore(): PluginConfigStore {
  const state = emptyState()
  return {
    readConfig: () => structuredClone(state.config),
    writeConfig: (id, config) => { touch(); if (config === undefined) delete state.config[id]; else state.config[id] = structuredClone(config) },
    readEnabled: () => structuredClone(state.enabled),
    writeEnabled: (id, enabled) => { touch(); state.enabled[id] = enabled },
    readQuarantine: () => structuredClone(state.quarantine),
    writeQuarantine: records => { touch(); state.quarantine = structuredClone(records) },
    readLaunchPlan: () => (state.launchPlan ? structuredClone(state.launchPlan) : null),
    writeLaunchPlan: plan => { touch(); state.launchPlan = structuredClone(plan) },
  }

  function touch(): void { state.updatedAt = new Date().toISOString() }
}

export interface FilePluginConfigStoreOptions {
  /** 读写失败时的兜底：例如文件损坏时以空状态起步而不抛出（Core 不可用时管理器必须仍可用）。 */
  onCorrupt?: (error: unknown) => void
}

export function createNodeFilePluginConfigStore(filePath: string, options: FilePluginConfigStoreOptions = {}): PluginConfigStore {
  const load = (): PersistedState => {
    try {
      const raw = readFileSync(filePath, 'utf8') as string
      const parsed = JSON.parse(raw) as Partial<PersistedState>
      if (!parsed || parsed.formatVersion !== STORE_FORMAT_VERSION) throw new Error(`unsupported store format: ${String(parsed?.formatVersion)}`)
      return {
        formatVersion: STORE_FORMAT_VERSION,
        updatedAt: parsed.updatedAt ?? new Date(0).toISOString(),
        config: parsed.config ?? {},
        enabled: parsed.enabled ?? {},
        quarantine: parsed.quarantine ?? [],
        launchPlan: parsed.launchPlan ?? null,
      }
    } catch (error) {
      options.onCorrupt?.(error)
      return emptyState()
    }
  }

  const persist = (state: PersistedState): void => {
    state.updatedAt = new Date().toISOString()
    // 原子替换：先写临时文件再 rename，避免中断产生半截 JSON。
    const temporary = `${filePath}.tmp`
    mkdirSync(dirname(filePath), { recursive: true })
    writeFileSync(temporary, JSON.stringify(state, null, 2), 'utf8')
    renameSync(temporary, filePath)
  }

  return {
    readConfig: () => structuredClone(load().config),
    writeConfig: (id, config) => { const state = load(); if (config === undefined) delete state.config[id]; else state.config[id] = structuredClone(config); persist(state) },
    readEnabled: () => structuredClone(load().enabled),
    writeEnabled: (id, enabled) => { const state = load(); state.enabled[id] = enabled; persist(state) },
    readQuarantine: () => structuredClone(load().quarantine),
    writeQuarantine: records => { const state = load(); state.quarantine = structuredClone(records); persist(state) },
    readLaunchPlan: () => load().launchPlan ? structuredClone(load().launchPlan!) : null,
    writeLaunchPlan: plan => { const state = load(); state.launchPlan = structuredClone(plan); persist(state) },
  }
}
